crates/bevy_asset/src/io/wasm_asset_io.rs

```diff
@@ -36,20 +36,34 @@ impl AssetIo for WasmAssetIo {
 
     fn read_directory(
         &self,
-        _path: &Path,
-    ) -> Result<Box<dyn Iterator<Item = PathBuf>>, AssetIoError> {
-        Ok(Box::new(std::iter::empty::<PathBuf>()))
+        _path: &str,
+    ) -> Result<Box<dyn Iterator<Item = String>>, AssetIoError> {
+        Ok(Box::new(std::iter::empty::<String>()))
     }
 
-    fn watch_path_for_changes(&self, _path: &Path) -> Result<(), AssetIoError> {
+    fn watch_path_for_changes(&self, _path: &str) -> Result<(), AssetIoError> {
         Ok(())
     }
 
     fn watch_for_changes(&self) -> Result<(), AssetIoError> {
         Ok(())
     }
 
-    fn is_directory(&self, path: &Path) -> bool {
+    fn is_directory(&self, path: &str) -> bool {
         self.root_path.join(path).is_dir()
     }
+
+    fn extension<'a>(&self, path: &'a str) -> Option<&'a str> {
```

Comment:
I'm not sure these belong in AssetIo. extension/hierarchy logic in a "path string" should be the same for every backend. Thats the main reason I'm using `Path` over `&str` right now. It does the "path logic" for us in a universal and "standard rust" way that people are used to.

I didn't highlight `is_directory` because that _should_ be an AssetIo behavior. Its unfortunate that `Path` has a few "filesystem" methods like `is_directory()`, which makes it an "impure" abstraction for what we're using it for.

In summary:
* We want asset paths to behave like Path in most cases in a backend-agnostic fashion
* Path has a couple of methods that make it impure that should live in AssetIo
* I'm happy to migrate to `&str`, provided we provide some sort of backend-agnostic path logic that can be used on `&str` without passing in state of some kind.

I would personally just continue to use Path instead of re-implementing the subset of logic we need for `&str`, but I'd accept this pr if we added something like `impl PathMethods for &str {}`.

I'd also like to point out that all of this code will likely get replaced by Atelier Assets at some point in the near-ish future.

Reply:
I disagree, I think an asset backend could be anything from a file system (the easy case) to a database (then the path could be an id) or http server (path could still be an actual path, or a query parameter, ...). In those case getting the extension could require asking the backend for it. Or even with file systems, assets could be stored by hash and splitted in several files for one asset if too large (not sure if this apply here, but for example at work we are using streamable data format that we read by chunk from the file system).

> I'd also like to point out that all of this code will likely get replaced by Atelier Assets at some point in the near-ish future.

I haven't wrapped my head around how Atelier Assets works yet, but hopefully it can still be backend independent...

Reply:
@mockersf  Haha I also disagree. Backends _could_ / should be those things, but the point is that the "path" is a common interface that _all_ backends support. It is the backend's job to translate between the "common" identifier and the "backend-specific" identifier.

ex: a database might store the "path" as the primary key for the asset. A webserver backend might translate a "texture.png" request to `example.com/assets/texture.png`.

People shouldn't need to rewrite game logic because they want to load assets from a webserver instead of a local file.

Atelier is making really good progress on being backend-independent (they recently added an asset io interface). They support loading assets by UUID (assigned by atelier), Path (retrieved from the filesystem), or other "custom" identifiers stored in the metadata.

Its a flexible system, so you could probably use alternative backend-specific ids if you wanted to. But I would not encourage that in Bevy. The whole point is that its "write once run anywhere" and I've selected Path to be the "common" key (as that is the interface people interact with when producing assets).

Reply:
I see your point, you would rather have the backends adapt to work uniformly than the user work only on one backend, that makes sense.

thanks for the feedback, closing now!